### ISSUE TITLE
snapshot waiting

### ISSUE THREAD

**Issue by impl1874** [caprica/vlcj#1028]
When the video playback is unstable, or disconnected for a few seconds, continue to play successfully,but,run snapshots method will be watting,It may be a problem with vlc,But here I want to ask for your help, thank you.

#my env
```
vlc              : 3.0.8
vlcj             : 4.7.0
os               : Ubuntu 18.04.5 LTS Linux 5.4.0-54-generic amd64
java             : 1.8.0_162 Oracle Corporation
java.home        : /usr/local/jdk/jdk1.8.0_162/jre
```

#my options
```
--no-osd"
--no-video-title-show"
--no-snapshot-preview"
--network-caching=1000"
--reset-config
--reset-plugins-cache
--snapshot-format=jpg
--vout=dummy
```

#my watting code line
```
BufferedImage image = mediaPlayer.snapshots().get()
```

#valcj watting code lines
```
file = File.createTempFile("vlcj-snapshot-", ".png");`
return ImageIO.read(new File(new WaitForSnapshot(mediaPlayer, file, width, height).await()));
```

#some vlc log
```
[00007f4ca441d380] main input error: ES_OUT_SET_(GROUP_)PCR  is called too late (pts_delay increased to 2056 ms)
[00007f4ca441d380] main input debug: ES_OUT_RESET_PCR called
[00007f4ca441d380] main input debug: Buffering 0%
**[00007f4cd40f1cb0] main video output error: Failed to grab a snapshot**
[NULL @ 0x7f4cbc0e9aa0] ct_type:0 pic_struct:0
[00007f4ca441d380] main input debug: Stream buffering done (9979 ms in 3046 ms)
[00007f4cbc152f30] main decoder error: buffer deadlock prevented
[00007f4ca441d380] main input debug: Decoder wait done in 0 ms
[h264 @ 0x7f4cbc889100] nal_unit_type: 7, nal_ref_idc: 3
[h264 @ 0x7f4cbc889100] nal_unit_type: 8, nal_ref_idc: 3
[h264 @ 0x7f4cbc889100] nal_unit_type: 6, nal_ref_idc: 0
[h264 @ 0x7f4cbc889100] nal_unit_type: 6, nal_ref_idc: 0
[h264 @ 0x7f4cbc889100] nal_unit_type: 5, nal_ref_idc: 3
[h264 @ 0x7f4cbc889100] ct_type:0 pic_struct:0
[00007f4cbc152f30] avcodec decoder debug: available hardware decoder output format 109 (vdpau)
[00007f4cbc152f30] avcodec decoder debug: available hardware decoder output format 53 (vaapi_vld)
[00007f4cbc152f30] avcodec decoder debug: available software decoder output format 0 (yuv420p)
[00007f4cbc152f30] avcodec decoder debug: reusing decoder output format 0
[h264 @ 0x7f4cbc889100] Reinit context to 1024x576, pix_fmt: yuv420p
```

**Comment by caprica** [comment]
According to the log, it looks like you are trying to take a snapshot while the video is buffering.

You should only take a snapshot if you have received a "playing" event for the current media.

vlcj could maybe handle this a bit better, perhaps by offering a timeout when it invokes await(). The underlying problem is that LibVLC does not send an event if the snapshot fails so in this situation the code will wait forever.

In short, change your code so it does not attempt a snapshot unless you know the media is playing (track that using events).

In the future, vlcj may be changed to support a timeout while waiting for the snapshot - so it would still fail, but at least it wouldn't block forever.

**Comment by impl1874** [comment]
thanks,have a good day.

**Comment by caprica** [comment]
The method you found does not wait/block, BUT when that method returns it is unlikely that a snapshot has been created _yet_.

That method is asynchronous, the snapshot is not "ready" until a subsequent snapshot-taken event is received, and the other blocking method waits for that.

You can of course implement your own wait mechanism, but simply using this method instead of the original one you used is not the full solution.
